

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Prepare and run imputation &#8212; scHiCluster</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-codeautolink.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Tan2021/imputation';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="scRNA(MALBAC) clustering" href="rna_clustering.html" />
    <link rel="prev" title="Tan et al. Cell 2021." href="intro.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    <p class="title logo__title">scHiCluster</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    scHiCluster
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorial</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="intro.html">Tan et al. Cell 2021.</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Prepare and run imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_clustering.html">scRNA(MALBAC) clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="dipc_clustering.html">DipC clustering at 100kb resolution</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_3c_integrate.html">Assign DipC cells to RNA clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="prepare.html">Prepare pseudobulk analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="loop_blacklist.html">Generate blacklist for loop calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../hba/intro.html">Human Brain Atlas</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../hba/rna_majortype/intro.html">Differentially expressed genes</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../hba/rna_majortype/01.deg.html">DEG identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/rna_majortype/02.loop_corr.html">RNA correlated loop</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/rna_majortype/03.domain_corr.html">RNA correlated boundary</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/rna_majortype/04.comp_corr.html">RNA correlated compartment</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../hba/loop_majortype/intro.html">Differential loops</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../hba/loop_majortype/01.diff.html">ANOVA of loop matrices across cell types</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/loop_majortype/02.load.html">Load loop matrices and stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/loop_majortype/03.apa.html">Aggregate peak analysis of differential loops</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/loop_majortype/04.corr.html">Correlation between loop and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/loop_majortype/05.plot.html">Plot correlation between loop and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../hba/domain_majortype/intro.html">Differential domain boundaries</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../hba/domain_majortype/01.call.html">Domain boundary prob</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/domain_majortype/02.count.html">Domain count and size</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/domain_majortype/03.diff.html">Differential boundary</a></li>

<li class="toctree-l3"><a class="reference internal" href="../hba/domain_majortype/04.corr.html">Correlation between boundary and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/domain_majortype/05.plot.html">Plot correlation between boundary and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../hba/compartment_majortype/intro.html">Differential compartments</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/01.call.html">Call compartment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/02.comC.html">Compare compartment with comethylation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/03.example.html">Plot example</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/04.diff.html">Differential compartment calling with dcHiC</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/05.corr.html">Correlation between compartment and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hba/compartment_majortype/06.plot.html">Plot correlation between compartment and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../hba/correlation_summary.html">Correlation Summary</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">COMMAND LINE TOOLS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../command_line/hicluster_filter_contact.html">hicluster filter-contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="../command_line/hicluster_prepare_impute.html">hicluster prepare-impute</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/zhoujt1994/scHiCluster" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/zhoujt1994/scHiCluster/issues/new?title=Issue%20on%20page%20%2FTan2021/imputation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Tan2021/imputation.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Prepare and run imputation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="prepare-and-run-imputation">
<h1>Prepare and run imputation<a class="headerlink" href="#prepare-and-run-imputation" title="Permalink to this heading">#</a></h1>
<p>We noticed strong signals at ENCODE blacklist regions that looks artificial. Therefore we first filter the chromatin contacts by the blacklist. In snm3C-seq data, we also provide a 2d blacklist to remove other fake signals.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate a contact file list</span>
ls<span class="w"> </span>raw/*/contacts_pairs/*.contacts.pairs.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">p</span><span class="o">=</span><span class="nv">$PWD</span><span class="w"> </span><span class="s1">&#39;{printf(&quot;%s/%s\n&quot;, p, $0)}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>contact_table.txt
paste<span class="w"> </span>&lt;<span class="o">(</span>awk<span class="w"> </span>-F<span class="s1">&#39;/&#39;</span><span class="w"> </span><span class="s1">&#39;{print $NF}&#39;</span><span class="w"> </span>contact_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d.<span class="w"> </span>-f1<span class="o">)</span><span class="w"> </span>contact_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>&gt;<span class="w"> </span>contact_table.tsv

<span class="c1"># Remove blacklist</span>
hicluster<span class="w"> </span>filter-contact<span class="w"> </span>--output_dir<span class="w"> </span>rmbkl/<span class="w"> </span>--blacklist_1d_path<span class="w"> </span>mm10-blacklist.v2.bed.gz<span class="w"> </span>--chr1<span class="w"> </span><span class="m">1</span><span class="w"> </span>--pos1<span class="w"> </span><span class="m">2</span><span class="w"> </span>--chr2<span class="w"> </span><span class="m">3</span><span class="w"> </span>--pos2<span class="w"> </span><span class="m">4</span><span class="w"> </span>--contact_table<span class="w"> </span>contact_table.tsv<span class="w"> </span>--chrom_size_path<span class="w"> </span>chrom_sizes.txt
</pre></div>
</div>
<p>The following command will generate a snakemake_cmd.txt file, where each line is a snakemake command to run imputation with certain parameters for a number of cells (batch_size). Thus, you can append each line to your job submission template to parallelize between batches. The parallelization between cells in the same batch will be handled by snakemake on a single node. You can decide the batch_size based on your super computer setting. In the example below, our imputation would be run on nodes using 96 cpus per node, and we assigned 1536 cells to a single job so that it would not create too many jobs to avoid long queue time.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate a contact file list</span>
ls<span class="w"> </span>rmbkl/*<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">p</span><span class="o">=</span><span class="nv">$PWD</span><span class="w"> </span><span class="s1">&#39;{printf(&quot;%s/%s\n&quot;, p, $0)}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>contact_table_rmbkl.txt
paste<span class="w"> </span>&lt;<span class="o">(</span>awk<span class="w"> </span>-F<span class="s1">&#39;/&#39;</span><span class="w"> </span><span class="s1">&#39;{print $NF}&#39;</span><span class="w"> </span>contact_table_rmbkl.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d.<span class="w"> </span>-f1<span class="o">)</span><span class="w"> </span>contact_table_rmbkl.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>&gt;<span class="w"> </span>contact_table_rmbkl.tsv

<span class="c1"># Prepare imputation </span>
<span class="c1"># Contact file list are divided into smaller chunks assigned to different jobs</span>
<span class="c1"># Commands to run imputation are generated</span>
hicluster<span class="w"> </span>prepare-impute<span class="w"> </span>--cell_table<span class="w"> </span>contact_table_rmbkl.tsv<span class="w"> </span>--batch_size<span class="w"> </span><span class="m">1536</span><span class="w"> </span>--pad<span class="w"> </span><span class="m">1</span><span class="w"> </span>--cpu_per_job<span class="w"> </span><span class="m">96</span><span class="w"> </span>--chr1<span class="w"> </span><span class="m">1</span><span class="w"> </span>--pos1<span class="w"> </span><span class="m">2</span><span class="w"> </span>--chr2<span class="w"> </span><span class="m">3</span><span class="w"> </span>--pos2<span class="w"> </span><span class="m">4</span><span class="w"> </span>--output_dir<span class="w"> </span>impute/100K/<span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--output_dist<span class="w"> </span><span class="m">500000000</span><span class="w"> </span>--window_size<span class="w"> </span><span class="m">500000000</span><span class="w"> </span>--step_size<span class="w"> </span><span class="m">500000000</span><span class="w"> </span>--resolution<span class="w"> </span><span class="m">100000</span>
hicluster<span class="w"> </span>prepare-impute<span class="w"> </span>--cell_table<span class="w"> </span>contact_table_rmbkl.tsv<span class="w"> </span>--batch_size<span class="w"> </span><span class="m">1536</span><span class="w"> </span>--pad<span class="w"> </span><span class="m">2</span><span class="w"> </span>--cpu_per_job<span class="w"> </span><span class="m">96</span><span class="w"> </span>--chr1<span class="w"> </span><span class="m">1</span><span class="w"> </span>--pos1<span class="w"> </span><span class="m">2</span><span class="w"> </span>--chr2<span class="w"> </span><span class="m">3</span><span class="w"> </span>--pos2<span class="w"> </span><span class="m">4</span><span class="w"> </span>--output_dir<span class="w"> </span>impute/25K/<span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--output_dist<span class="w"> </span><span class="m">5050000</span><span class="w"> </span>--window_size<span class="w"> </span><span class="m">500000000</span><span class="w"> </span>--step_size<span class="w"> </span><span class="m">500000000</span><span class="w"> </span>--resolution<span class="w"> </span><span class="m">25000</span>
hicluster<span class="w"> </span>prepare-impute<span class="w"> </span>--cell_table<span class="w"> </span>contact_table_rmbkl.tsv<span class="w"> </span>--batch_size<span class="w"> </span><span class="m">1536</span><span class="w"> </span>--pad<span class="w"> </span><span class="m">2</span><span class="w"> </span>--cpu_per_job<span class="w"> </span><span class="m">96</span><span class="w"> </span>--chr1<span class="w"> </span><span class="m">1</span><span class="w"> </span>--pos1<span class="w"> </span><span class="m">2</span><span class="w"> </span>--chr2<span class="w"> </span><span class="m">3</span><span class="w"> </span>--pos2<span class="w"> </span><span class="m">4</span><span class="w"> </span>--output_dir<span class="w"> </span>impute/10K/<span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--output_dist<span class="w"> </span><span class="m">5050000</span><span class="w"> </span>--window_size<span class="w"> </span><span class="m">30000000</span><span class="w"> </span>--step_size<span class="w"> </span><span class="m">10000000</span><span class="w"> </span>--resolution<span class="w"> </span><span class="m">10000</span>
mkdir<span class="w"> </span>-p<span class="w"> </span>sbatch/impute/
cat<span class="w"> </span>impute/*/snakemake_cmd.txt<span class="w"> </span>&gt;<span class="w"> </span>sbatch/impute/snakemake_cmd.txt
</pre></div>
</div>
<p>Then append each line of snakemake_cmd.txt to your job submission template. Submit the jobs and run imputation! With our setting (standard node on <a class="reference external" href="http://anvil.rcac.purdue.edu">anvil.rcac.purdue.edu</a>), imputation of 1536 cells at 100kb resolution takes 30 minutes, 25kb resolution takes 2.5 hours, and 10kb resolution takes 5 hours.</p>
<p>After Imputation, you can use the following commands to generate single-cell embedding and several cell-by-feature matrices, including compartment score matrix, insulation score matrix, domain boundary matrix, and gene interaction strength matrix. These commands are also resource consuming, so please consider submitting them to computing nodes. With the same setting as described above, embedding takes ~8 minutes, compartment with raw matrices takes ~1 minutes, with imputed matrices takes ~2 minutes, domain takes ~75 minutes, gene score takes ~15 minutes.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute embedding</span>
ls<span class="w"> </span>impute/100K/chunk*/*.cool<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">p</span><span class="o">=</span><span class="nv">$PWD</span><span class="w"> </span><span class="s1">&#39;{printf(&quot;%s/%s\n&quot;, p, $0)}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>impute/100K/cell_table.txt
paste<span class="w"> </span>&lt;<span class="o">(</span>awk<span class="w"> </span>-F<span class="s1">&#39;/&#39;</span><span class="w"> </span><span class="s1">&#39;{print $NF}&#39;</span><span class="w"> </span>impute/100K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d.<span class="w"> </span>-f1<span class="o">)</span><span class="w"> </span>impute/100K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>&gt;<span class="w"> </span>impute/100K/cell_table.tsv
hicluster<span class="w"> </span>embedding<span class="w"> </span>--cell_table_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/impute/100K/cell_table.tsv<span class="w"> </span>--output_dir<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/dataset/embedding<span class="w"> </span>--dim<span class="w"> </span><span class="m">50</span><span class="w"> </span>--dist<span class="w"> </span><span class="m">1000000</span><span class="w"> </span>--resolution<span class="w"> </span><span class="m">100000</span><span class="w"> </span>--scale_factor<span class="w"> </span><span class="m">100000</span><span class="w"> </span>--norm_sig<span class="w"> </span>--save_raw<span class="w"> </span>--cpu<span class="w"> </span><span class="m">20</span>

<span class="c1"># Compute compartment score using raw contact matrices</span>
<span class="c1">## CpG density</span>
hicluster<span class="w"> </span>cpg-ratio<span class="w"> </span>--fasta_path<span class="w"> </span>/anvil/projects/x-mcb130189/mm10/mm10_with_chrl.fa<span class="w"> </span>--hdf_output_path<span class="w"> </span>cpg_ratio_100k.hdf<span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--resolution<span class="w"> </span><span class="m">100000</span>
<span class="c1">## Using raw matrices</span>
hicluster<span class="w"> </span>compartment<span class="w"> </span>--cell_table_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/contact_table_rmbkl.tsv<span class="w"> </span>--output_prefix<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/dataset/Tan2021.raw<span class="w"> </span>--cpg_profile_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/cpg_ratio_100k.hdf<span class="w"> </span>--cpu<span class="w"> </span><span class="m">96</span><span class="w"> </span>--resolution<span class="w"> </span><span class="m">100000</span><span class="w"> </span>--chr1<span class="w"> </span><span class="m">1</span><span class="w"> </span>--pos1<span class="w"> </span><span class="m">2</span><span class="w"> </span>--chr2<span class="w"> </span><span class="m">3</span><span class="w"> </span>--pos2<span class="w"> </span><span class="m">4</span><span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--mode<span class="w"> </span>tsv
<span class="c1">## Using imputed matrices</span>
hicluster<span class="w"> </span>compartment<span class="w"> </span>--cell_table_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/impute/100K/cell_table.tsv<span class="w"> </span>--output_prefix<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/dataset/Tan2021.impute<span class="w"> </span>--cpg_profile_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/cpg_ratio_100k.hdf<span class="w"> </span>--cpu<span class="w"> </span><span class="m">96</span>

<span class="c1"># Compute insulation score and domain boundary</span>
ls<span class="w"> </span>impute/25K/chunk*/*.cool<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">p</span><span class="o">=</span><span class="nv">$PWD</span><span class="w"> </span><span class="s1">&#39;{printf(&quot;%s/%s\n&quot;, p, $0)}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>impute/25K/cell_table.txt
paste<span class="w"> </span>&lt;<span class="o">(</span>awk<span class="w"> </span>-F<span class="s1">&#39;/&#39;</span><span class="w"> </span><span class="s1">&#39;{print $NF}&#39;</span><span class="w"> </span>impute/25K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d.<span class="w"> </span>-f1<span class="o">)</span><span class="w"> </span>impute/25K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>&gt;<span class="w"> </span>impute/25K/cell_table.tsv
hicluster<span class="w"> </span>domain<span class="w"> </span>--cell_table_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/impute/25K/cell_table.tsv<span class="w"> </span>--output_prefix<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/dataset/Tan2021<span class="w"> </span>--resolution<span class="w"> </span><span class="m">25000</span><span class="w"> </span>--window_size<span class="w"> </span><span class="m">10</span><span class="w"> </span>--cpu<span class="w"> </span><span class="m">96</span>

<span class="c1"># Compute gene contact score</span>
ls<span class="w"> </span>impute/10K/chunk*/*.cool<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">p</span><span class="o">=</span><span class="nv">$PWD</span><span class="w"> </span><span class="s1">&#39;{printf(&quot;%s/%s\n&quot;, p, $0)}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>impute/10K/cell_table.txt
paste<span class="w"> </span>&lt;<span class="o">(</span>awk<span class="w"> </span>-F<span class="s1">&#39;/&#39;</span><span class="w"> </span><span class="s1">&#39;{print $NF}&#39;</span><span class="w"> </span>impute/10K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d.<span class="w"> </span>-f1<span class="o">)</span><span class="w"> </span>impute/10K/cell_table.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>&gt;<span class="w"> </span>impute/10K/cell_table.tsv
hicluster<span class="w"> </span>gene-score<span class="w"> </span>--cell_table_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/impute/10K/cell_table.tsv<span class="w"> </span>--gene_meta_path<span class="w"> </span>/anvil/projects/x-mcb130189/gene/vm22/gencode.vM22.annotation.gene.sorted.bed.gz<span class="w"> </span>--resolution<span class="w"> </span><span class="m">10000</span><span class="w"> </span>--output_hdf_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/dataset/Tan2021.geneimputescore.hdf<span class="w"> </span>--chrom_size_path<span class="w"> </span>/anvil/scratch/x-zhou/Tan2021/chrom_sizes.txt<span class="w"> </span>--cpu<span class="w"> </span><span class="m">96</span><span class="w"> </span>--mode<span class="w"> </span>impute
</pre></div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Tan2021"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="intro.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Tan et al. Cell 2021.</p>
      </div>
    </a>
    <a class="right-next"
       href="rna_clustering.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">scRNA(MALBAC) clustering</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Jingtian Zhou
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      Â© Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>