

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>schicluster.loop.merge_raw_matrix &#8212; scHiCluster</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-codeautolink.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/schicluster/loop/merge_raw_matrix';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../intro.html">
  
  
  
  
  
    <p class="title logo__title">scHiCluster</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../intro.html">
                    scHiCluster
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorial</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../Tan2021/intro.html">Tan et al. Cell 2021.</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/imputation.html">Prepare and run imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/rna_clustering.html">scRNA(MALBAC) clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/dipc_clustering.html">DipC clustering at 100kb resolution</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/rna_3c_integrate.html">Assign DipC cells to RNA clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/prepare.html">Prepare pseudobulk analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Tan2021/loop_blacklist.html">Generate blacklist for loop calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../hba/intro.html">Human Brain Atlas</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../hba/rna_majortype/intro.html">Differentially expressed genes</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/rna_majortype/01.deg.html">DEG identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/rna_majortype/02.loop_corr.html">RNA correlated loop</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/rna_majortype/03.domain_corr.html">RNA correlated boundary</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/rna_majortype/04.comp_corr.html">RNA correlated compartment</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../hba/loop_majortype/intro.html">Differential loops</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/loop_majortype/01.diff.html">ANOVA of loop matrices across cell types</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/loop_majortype/02.load.html">Load loop matrices and stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/loop_majortype/03.apa.html">Aggregate peak analysis of differential loops</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/loop_majortype/04.corr.html">Correlation between loop and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/loop_majortype/05.plot.html">Plot correlation between loop and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../hba/domain_majortype/intro.html">Differential domain boundaries</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/domain_majortype/01.call.html">Domain boundary prob</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/domain_majortype/02.count.html">Domain count and size</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/domain_majortype/03.diff.html">Differential boundary</a></li>

<li class="toctree-l3"><a class="reference internal" href="../../../hba/domain_majortype/04.corr.html">Correlation between boundary and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/domain_majortype/05.plot.html">Plot correlation between boundary and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../hba/compartment_majortype/intro.html">Differential compartments</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/01.call.html">Call compartment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/02.comC.html">Compare compartment with comethylation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/03.example.html">Plot example</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/04.diff.html">Differential compartment calling with dcHiC</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/05.corr.html">Correlation between compartment and other modalities</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../hba/compartment_majortype/06.plot.html">Plot correlation between compartment and mC/ATAC/RNA</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../hba/correlation_summary.html">Correlation Summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../hba/plot_example.html">Plot examples</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">COMMAND LINE TOOLS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_filter_contact.html">hicluster filter-contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_prepare_impute.html">hicluster prepare-impute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_contact_distance.html">hicluster contact-distance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_embedding.html">hicluster embedding</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_cpg_ratio.html">hicluster cpg-ratio</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_compartment.html">hicluster compartment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_domain.html">hicluster domain</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_gene_score.html">hicluster gene-score</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_merge_cell_raw.html">hicluster merge-cell-raw</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line/hicluster_merge_cool.html">hicluster merge-cool</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/zhoujt1994/scHiCluster" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/zhoujt1994/scHiCluster/issues/new?title=Issue%20on%20page%20%2F_modules/schicluster/loop/merge_raw_matrix.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <h1>Source code for schicluster.loop.merge_raw_matrix</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">cooler</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">from</span> <span class="nn">concurrent.futures</span> <span class="kn">import</span> <span class="n">ProcessPoolExecutor</span><span class="p">,</span> <span class="n">as_completed</span>

<span class="kn">from</span> <span class="nn">..cool</span> <span class="kn">import</span> <span class="n">get_chrom_offsets</span>
<span class="kn">from</span> <span class="nn">.merge_cell_to_group</span> <span class="kn">import</span> <span class="n">read_single_cool_chrom</span>


<div class="viewcode-block" id="_chrom_sum_iterator"><a class="viewcode-back" href="../../../api/schicluster/loop/merge_raw_matrix/index.html#schicluster.loop.merge_raw_matrix._chrom_sum_iterator">[docs]</a><span class="k">def</span> <span class="nf">_chrom_sum_iterator</span><span class="p">(</span><span class="n">cell_urls</span><span class="p">,</span>
                        <span class="n">chrom_sizes</span><span class="p">,</span>
                        <span class="n">chrom_offset</span><span class="p">,</span>
                        <span class="n">add_trans</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Iterate through the raw matrices and chromosomes of cells.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cell_urls :</span>
<span class="sd">        List of cell urls.</span>
<span class="sd">    chrom_sizes :</span>
<span class="sd">        Dictionary of chromosome sizes.</span>
<span class="sd">    chrom_offset :</span>
<span class="sd">        Dictionary of chromosome offsets.</span>
<span class="sd">    add_trans :</span>
<span class="sd">        If true, will also iterate all the trans combinations (different chromosomes).</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    pixel_df :</span>
<span class="sd">        Dataframe of pixels. Used by cooler.create_cooler to save to h5 file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_iter_1d</span><span class="p">(</span><span class="n">_chrom1</span><span class="p">,</span> <span class="n">_chrom2</span><span class="p">):</span>
        <span class="c1"># sum together multiple chunks</span>
        <span class="c1"># first</span>
        <span class="n">cell_url</span> <span class="o">=</span> <span class="n">cell_urls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">matrix</span> <span class="o">=</span> <span class="n">read_single_cool_chrom</span><span class="p">(</span><span class="n">cell_url</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="n">_chrom1</span><span class="p">,</span> <span class="n">chrom2</span><span class="o">=</span><span class="n">_chrom2</span><span class="p">)</span>

        <span class="c1"># others</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_urls</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">cell_url</span> <span class="ow">in</span> <span class="n">cell_urls</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">matrix</span> <span class="o">+=</span> <span class="n">read_single_cool_chrom</span><span class="p">(</span><span class="n">cell_url</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="n">_chrom1</span><span class="p">,</span> <span class="n">chrom2</span><span class="o">=</span><span class="n">_chrom2</span><span class="p">)</span>

        <span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">tocoo</span><span class="p">()</span>
        <span class="n">_pixel_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;bin1_id&#39;</span><span class="p">:</span> <span class="n">matrix</span><span class="o">.</span><span class="n">row</span><span class="p">,</span>
            <span class="s1">&#39;bin2_id&#39;</span><span class="p">:</span> <span class="n">matrix</span><span class="o">.</span><span class="n">col</span><span class="p">,</span>
            <span class="s1">&#39;count&#39;</span><span class="p">:</span> <span class="n">matrix</span><span class="o">.</span><span class="n">data</span>
        <span class="p">})</span>
        <span class="k">if</span> <span class="n">_chrom2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># both row and col are chrom1</span>
            <span class="n">_pixel_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">+=</span> <span class="n">chrom_offset</span><span class="p">[</span><span class="n">_chrom1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># row is chrom1, add chrom1 offset</span>
            <span class="n">_pixel_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="n">chrom_offset</span><span class="p">[</span><span class="n">_chrom1</span><span class="p">]</span>
            <span class="c1"># col is chrom2, add chrom2 offset</span>
            <span class="n">_pixel_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">chrom_offset</span><span class="p">[</span><span class="n">_chrom2</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">_pixel_df</span>

    <span class="k">if</span> <span class="n">add_trans</span><span class="p">:</span>
        <span class="c1"># only iter upper triangle</span>
        <span class="c1"># chrom order by offset, small to large</span>
        <span class="n">chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_offset</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="n">n_chroms</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chroms</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_chroms</span><span class="p">):</span>
            <span class="n">chrom1</span> <span class="o">=</span> <span class="n">chroms</span><span class="p">[</span><span class="n">a</span><span class="p">]</span>
            <span class="n">chrom1_dfs</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">n_chroms</span><span class="p">):</span>
                <span class="n">chrom2</span> <span class="o">=</span> <span class="n">chroms</span><span class="p">[</span><span class="n">b</span><span class="p">]</span>
                <span class="n">pixel_df</span> <span class="o">=</span> <span class="n">_iter_1d</span><span class="p">(</span><span class="n">chrom1</span><span class="p">,</span> <span class="n">chrom2</span><span class="p">)</span>
                <span class="n">chrom1_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pixel_df</span><span class="p">)</span>
            <span class="n">chrom1_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">chrom1_dfs</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;bin1_id&#39;</span><span class="p">,</span> <span class="s1">&#39;bin2_id&#39;</span><span class="p">])</span>
            <span class="k">yield</span> <span class="n">chrom1_df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chrom_sizes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">pixel_df</span> <span class="o">=</span> <span class="n">_iter_1d</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">yield</span> <span class="n">pixel_df</span></div>


<div class="viewcode-block" id="_save_single_matrix_type"><a class="viewcode-back" href="../../../api/schicluster/loop/merge_raw_matrix/index.html#schicluster.loop.merge_raw_matrix._save_single_matrix_type">[docs]</a><span class="k">def</span> <span class="nf">_save_single_matrix_type</span><span class="p">(</span><span class="n">cooler_path</span><span class="p">,</span>
                             <span class="n">bins_df</span><span class="p">,</span>
                             <span class="n">cell_urls</span><span class="p">,</span>
                             <span class="n">chrom_sizes</span><span class="p">,</span>
                             <span class="n">chrom_offset</span><span class="p">,</span>
                             <span class="n">add_trans</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save a single matrix type Cool file from merging multiple cell urls.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cooler_path :</span>
<span class="sd">        Path to the output cool file.</span>
<span class="sd">    bins_df :</span>
<span class="sd">        Dataframe of bins. Created from chromosome sizes and resolution.</span>
<span class="sd">    cell_urls :</span>
<span class="sd">        List of cell urls to merge.</span>
<span class="sd">    chrom_sizes :</span>
<span class="sd">        Dictionary of chromosome sizes.</span>
<span class="sd">    chrom_offset :</span>
<span class="sd">        Dictionary of chromosome offsets.</span>
<span class="sd">    add_trans :</span>
<span class="sd">        Whether add trans matrix also.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chrom_iter</span> <span class="o">=</span> <span class="n">_chrom_sum_iterator</span><span class="p">(</span><span class="n">cell_urls</span><span class="p">,</span>
                                     <span class="n">chrom_sizes</span><span class="p">,</span>
                                     <span class="n">chrom_offset</span><span class="p">,</span>
                                     <span class="n">add_trans</span><span class="o">=</span><span class="n">add_trans</span><span class="p">)</span>
    <span class="n">cooler</span><span class="o">.</span><span class="n">create_cooler</span><span class="p">(</span><span class="n">cool_uri</span><span class="o">=</span><span class="n">cooler_path</span><span class="p">,</span>
                         <span class="n">bins</span><span class="o">=</span><span class="n">bins_df</span><span class="p">,</span>
                         <span class="n">pixels</span><span class="o">=</span><span class="n">chrom_iter</span><span class="p">,</span>
                         <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                         <span class="n">dtypes</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">})</span>
    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">cooler_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;group_n_cells&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_urls</span><span class="p">)</span>
    <span class="k">return</span></div>


<div class="viewcode-block" id="make_raw_matrix_cell_table"><a class="viewcode-back" href="../../../api/schicluster/loop/merge_raw_matrix/index.html#schicluster.loop.merge_raw_matrix.make_raw_matrix_cell_table">[docs]</a><span class="k">def</span> <span class="nf">make_raw_matrix_cell_table</span><span class="p">(</span><span class="n">cell_table</span><span class="p">,</span> <span class="n">resolution_str</span><span class="o">=</span><span class="s1">&#39;10K&#39;</span><span class="p">):</span>
    <span class="c1"># get all the raw matrix cell url automatically if the dir path is the default</span>
    <span class="n">scool_dirs</span> <span class="o">=</span> <span class="n">cell_table</span><span class="p">[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]))</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">cell_urls</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">scool_file_pattern</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;raw/*.</span><span class="si">{</span><span class="n">resolution_str</span><span class="si">}</span><span class="s1">.scool&#39;</span>
    <span class="k">for</span> <span class="n">scool_dir</span> <span class="ow">in</span> <span class="n">scool_dirs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">scool_path</span> <span class="ow">in</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">scool_dir</span><span class="p">)</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">scool_file_pattern</span><span class="p">):</span>
            <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">scool_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_cool</span><span class="p">:</span>
                <span class="n">cell_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">_cool</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
                <span class="k">for</span> <span class="n">cell_id</span> <span class="ow">in</span> <span class="n">cell_ids</span><span class="p">:</span>
                    <span class="n">cell_urls</span><span class="p">[</span><span class="n">cell_id</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">scool_path</span><span class="si">}</span><span class="s1">::/cells/</span><span class="si">{</span><span class="n">cell_id</span><span class="si">}</span><span class="s1">&#39;</span>
    <span class="n">raw_url_series</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">cell_urls</span><span class="p">)</span>

    <span class="c1"># delete old url</span>
    <span class="k">del</span> <span class="n">cell_table</span><span class="p">[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">]</span>
    <span class="c1"># add new url</span>
    <span class="n">cell_table</span><span class="p">[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">raw_url_series</span>

    <span class="n">na_cells</span> <span class="o">=</span> <span class="n">cell_table</span><span class="p">[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">na_cells</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">na_cells</span><span class="si">}</span><span class="s1"> cells do not have raw matrix.&#39;</span><span class="p">)</span>

    <span class="n">cell_table</span> <span class="o">=</span> <span class="n">cell_table</span><span class="p">[[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_group&#39;</span><span class="p">]]</span>
    <span class="k">return</span> <span class="n">cell_table</span></div>


<div class="viewcode-block" id="merge_raw_scool_by_cluster"><a class="viewcode-back" href="../../../api/schicluster/loop/merge_raw_matrix/index.html#schicluster.loop.merge_raw_matrix.merge_raw_scool_by_cluster">[docs]</a><span class="k">def</span> <span class="nf">merge_raw_scool_by_cluster</span><span class="p">(</span><span class="n">chrom_size_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">cell_table_path</span><span class="p">,</span>
                               <span class="n">output_dir</span><span class="p">,</span> <span class="n">add_trans</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sum the raw matrix of cells, no normalization.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chrom_size_path :</span>
<span class="sd">        Path to the chrom size file. This file is used to determine chromosome names and bins.</span>
<span class="sd">    resolution :</span>
<span class="sd">        Resolution of the raw matrix.</span>
<span class="sd">    cell_table_path :</span>
<span class="sd">        Path to the cell table.</span>
<span class="sd">        This table should contain three columns: cell_id, cell_url, cell_group; no Header.</span>
<span class="sd">        The cell_id is the id of the cell in the raw matrix.</span>
<span class="sd">        The cell_url is the path to the raw matrix.</span>
<span class="sd">        The cell_group is the group of the cells.</span>
<span class="sd">    output_dir :</span>
<span class="sd">        Path to the output directory. Group cool files will be named as &quot;&lt;output_dir&gt;/&lt;cell_group&gt;.cool&quot;.</span>
<span class="sd">    add_trans :</span>
<span class="sd">        Whether add trans matrix also.</span>
<span class="sd">    cpu :</span>
<span class="sd">        Number of CPUs to use.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># determine chunk dirs for the group:</span>
    <span class="n">output_dir</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
    <span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">cell_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cell_table_path</span><span class="p">,</span>
                             <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
                             <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                             <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                             <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_url&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_group&#39;</span><span class="p">])</span>

    <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">read_chromsizes</span><span class="p">(</span><span class="n">chrom_size_path</span><span class="p">,</span> <span class="n">all_names</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">bins_df</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">binnify</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    <span class="n">chrom_offset</span> <span class="o">=</span> <span class="n">get_chrom_offsets</span><span class="p">(</span><span class="n">bins_df</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">ProcessPoolExecutor</span><span class="p">(</span><span class="n">cpu</span><span class="p">)</span> <span class="k">as</span> <span class="n">exe</span><span class="p">:</span>
        <span class="n">futures</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">cell_group</span><span class="p">,</span> <span class="n">sub_df</span> <span class="ow">in</span> <span class="n">cell_table</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;cell_group&#39;</span><span class="p">):</span>
            <span class="n">cell_urls</span> <span class="o">=</span> <span class="n">sub_df</span><span class="p">[</span><span class="s1">&#39;cell_url&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="n">cooler_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cell_group</span><span class="si">}</span><span class="s1">.cool&#39;</span>
            <span class="k">if</span> <span class="n">cooler_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cooler_path</span><span class="si">}</span><span class="s1"> already exists, skip.&#39;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="n">cooler_temp_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cell_group</span><span class="si">}</span><span class="s1">.temp.cool&#39;</span><span class="p">)</span>
            <span class="n">future</span> <span class="o">=</span> <span class="n">exe</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="n">_save_single_matrix_type</span><span class="p">,</span>
                                <span class="n">cooler_path</span><span class="o">=</span><span class="n">cooler_temp_path</span><span class="p">,</span>
                                <span class="n">bins_df</span><span class="o">=</span><span class="n">bins_df</span><span class="p">,</span>
                                <span class="n">cell_urls</span><span class="o">=</span><span class="n">cell_urls</span><span class="p">,</span>
                                <span class="n">chrom_sizes</span><span class="o">=</span><span class="n">chrom_sizes</span><span class="p">,</span>
                                <span class="n">chrom_offset</span><span class="o">=</span><span class="n">chrom_offset</span><span class="p">,</span>
                                <span class="n">add_trans</span><span class="o">=</span><span class="n">add_trans</span><span class="p">)</span>
            <span class="n">futures</span><span class="p">[</span><span class="n">future</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_group</span>
        <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">futures</span><span class="p">):</span>
            <span class="n">cell_group</span> <span class="o">=</span> <span class="n">futures</span><span class="p">[</span><span class="n">future</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Matrix </span><span class="si">{</span><span class="n">cell_group</span><span class="si">}</span><span class="s1"> generated&#39;</span><span class="p">)</span>
            <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>

            <span class="c1"># move the temp cool file to the final location</span>
            <span class="n">cooler_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cell_group</span><span class="si">}</span><span class="s1">.cool&#39;</span><span class="p">)</span>
            <span class="n">cooler_temp_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cell_group</span><span class="si">}</span><span class="s1">.temp.cool&#39;</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">cooler_temp_path</span><span class="p">,</span> <span class="n">cooler_path</span><span class="p">)</span>
    <span class="k">return</span></div>
</pre></div>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Jingtian Zhou
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>